### PR TITLE
Fix scaffold to support rest-json style API

### PR DESCRIPTION
**Pull request by toshitanian**
I fixed scaffold script and `moto/core/responses.py` to support rest-json style API in scaffold.

# changes in core module
In `moto/core/responses.py`,  I mainly changed `_get_param` method and `_dispatch`+`_get_action` method.

**In `_get_param`**
Response class has been getting parameters only from queryparameter. In the PR, it will try to get parameters in order of queryparameter, json body and path parameter.

**In `_dispatch`+`_get_action`**
previously Response class had been dispatching request to appropriate method  based on the operation name in the request. Unfortunately, rest-api request does not include operation name in itself. In this PR, when it cannot find operation name in the request, it looks up operation name in boto3 client by request method and request uri.

# changes in scaffold
some refactors + fixes to adapt to change in core module for rest-api

### CODE REVIEW

**Comment by toshitanian** [comment]
okay, please review this ;)

**Comment by toshitanian** [comment]
@JackDanger This PR is ready. Can you review and merge this?

File: moto/core/responses.py [hunk 1]
```diff
@@ -178,6 +181,58 @@ def _dispatch(self, request, full_url, headers):
...
+        return regexp
+
+    def _get_action_from_method_and_request_uri(self, method, request_uri):
+        """basically used for `rest-json` APIs
+        You can refer to example from link below
+        https://github.com/boto/botocore/blob/develop/botocore/data/iot/2015-05-28/service-2.json
+        """
+
+        # service response class should have 'SERVICE_NAME' class member,
+        # if want to get action from method and url
```
Comment (JackDanger) [inline moto/core/responses.py]: Small grammar change in comment: `if want to get action` -> `if you want to get the action`

File: moto/core/responses.py [hunk 2]
```diff
@@ -178,6 +181,58 @@ def _dispatch(self, request, full_url, headers):
...
+        if not hasattr(self, 'SERVICE_NAME'):
+            return None
+        service = self.SERVICE_NAME
+        conn = boto3.client(service)
+
+        # make cache if it has not exist yet
```
Comment (JackDanger) [inline moto/core/responses.py]: I suggest changing to "make cache if it does not exist yet"

File: moto/core/responses.py [hunk 3]
```diff
@@ -221,6 +278,22 @@ def _get_param(self, param_name, if_none=None):
         val = self.querystring.get(param_name)
         if val is not None:
             return val[0]
+
+        # try to get json body parameter
+        try:
+            j = json.loads(self.body)
+            # raise key error if key really does not exist
+            return j[param_name]
+        except:
```
Comment (JackDanger) [inline moto/core/responses.py]: This `except` catches a bit too much. How about we add clauses for `except JSONDecodeError` and `except KeyError` for the two things that could go wrong here?

```
try:
  return json.loads(self.body)[param_name]
except JSONDecodeError: pass
except KeyError: pass
```

File: moto/core/responses.py [hunk 4]
```diff
@@ -221,6 +278,22 @@ def _get_param(self, param_name, if_none=None):
...
+            # do nothing if param is not found
+            pass
+        # try to get path parameter
+        if self.uri_match:
+            try:
+                return self.uri_match.group(param_name)
+            except:
```
Comment (JackDanger) [inline moto/core/responses.py]: How about `except IndexError:` instead?

File: scripts/scaffold.py [hunk 5]
```diff
@@ -413,81 +429,55 @@ def insert_url(service, operation):
...
-
-def insert_json_codes(service, operation):
-    func_in_responses = get_function_in_responses(service, operation, 'json')
-    func_in_models = get_function_in_models(service, operation)
-
-    # edit responses.py
-    responses_path = 'moto/{}/responses.py'.format(service)
-    print_progress('inserting code', responses_path, 'green')
-    insert_code_to_class(responses_path, BaseResponse, func_in_responses)
+    if api_protocol == 'query':
```
Comment (JackDanger) [inline scripts/scaffold.py]: Oooh, nice!

File: moto/core/responses.py (hunk 3 above)
Comment (toshitanian) [reply moto/core/responses.py]: thanks, I will use Value error instead of JSONDecodeError to support both python2 and python3

Comment (toshitanian) [reply moto/core/responses.py]: And, made new line for each exception handling because linter warns.

**Comment by toshitanian** [comment]
Thank you for reviewing for your early morning ;) I fixed the points you mentioned. @JackDanger

**Comment by toshitanian** [comment]
sorry, ci is still running....

**Comment by JackDanger** [comment]
Thanks, @toshitanian! Great work on this, I'm so glad we have better tools for the people working inside Moto.

The builds are a bit backed up but once this passes I'll merge it.

**Comment by toshitanian** [comment]
🤔

**Comment by toshitanian** [comment]
It seems that there's a bug for python2

**Comment by toshitanian** [comment]
🎉

**Pull request merged**